[Backport] Fixed Issue : Search REST API returns wrong total_count

ronak2ram opened the pull request:
### Original Pull Request 
 https://github.com/magento/magento2/pull/21713


### Description (*)
I added function for getting total_count of result item.

### Fixed Issues (if relevant)
1. magento/magento2#17295: Search REST API returns wrong total_count

### Manual testing scenarios (*)
### Request :
 `rest/V1/search?searchCriteria[requestName]=quick_search_container&searchCriteria[filterGroups][0][filters][0][field]=search_term&searchCriteria[filterGroups][0][filters][0][conditionType]=like&searchCriteria[filterGroups][0][filters][0][value]=test&searchCriteria[current_page]=1&searchCriteria[page_size]=2`

### Response : 
```
{
    "items": [
        {
            "id": 126,
            "custom_attributes": [
                {
                    "attribute_code": "score",
                    "value": "73.7660923004150400"
                }
            ]
        },
        {
            "id": 61,
            "custom_attributes": [
                {
                    "attribute_code": "score",
                    "value": "73.7660923004150400"
                }
            ]
        }
    ],
    "aggregations": {
        "buckets": [
            {
                "name": "category_bucket",
                "values": [
                    {
                        "value": "32",
                        "metrics": [
                            "32",
                            "1"
                        ]
                    }
                ]
            },
            {
                "name": "stock_status_bucket",
                "values": []
            },
            {
                "name": "rating_summary_bucket",
                "values": []
            },
            {
                "name": "am_is_new_bucket",
                "values": []
            },
            {
                "name": "am_on_sale_bucket",
                "values": []
            },
            {
                "name": "color_bucket",
                "values": []
            },
            {
                "name": "sub_category_bucket",
                "values": []
            }
        ],
        "bucket_names": [
            "category_bucket",
            "stock_status_bucket",
            "rating_summary_bucket",
            "am_is_new_bucket",
            "am_on_sale_bucket",
            "color_bucket",
            "sub_category_bucket"
        ]
    },
    "search_criteria": {
        "request_name": "quick_search_container",
        "filter_groups": [
            {
                "filters": [
                    {
                        "field": "search_term",
                        "value": "test",
                        "condition_type": "like"
                    }
                ]
            }
        ],
        "page_size": 2,
        "current_page": 1
    },
    "total_count": 9
}
```
### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

ronak2ram committed:
Resolve Issue : Search REST API returns wrong total_count

magento-engcom-team commented:
Hi @ronak2ram. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.2-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

dmytro-ch commented:
Hi @ronak2ram,
thank you for your contribution!

There are a couple of errors in unit tests with the current solution.

```
1) Magento\Framework\Search\Test\Unit\Adapter\Mysql\AdapterTest::testQuery
Exception: Notice: Undefined index: total in /home/travis/build/magento/magento2/lib/internal/Magento/Framework/Search/Test/Unit/Adapter/Mysql/AdapterTest.php on line 171
/home/travis/build/magento/magento2/lib/internal/Magento/Framework/App/ErrorHandler.php:61
/home/travis/build/magento/magento2/lib/internal/Magento/Framework/Search/Test/Unit/Adapter/Mysql/AdapterTest.php:171
2) Magento\Framework\Search\Test\Unit\Adapter\Mysql\ResponseFactoryTest::testCreate
Exception: Notice: Undefined index: total in /home/travis/build/magento/magento2/lib/internal/Magento/Framework/Search/Adapter/Mysql/ResponseFactory.php on line 69
/home/travis/build/magento/magento2/lib/internal/Magento/Framework/App/ErrorHandler.php:61
/home/travis/build/magento/magento2/lib/internal/Magento/Framework/Search/Adapter/Mysql/ResponseFactory.php:69
/home/travis/build/magento/magento2/lib/internal/Magento/Framework/Search/Test/Unit/Adapter/Mysql/ResponseFactoryTest.php:68
3) Magento\Framework\Search\Test\Unit\SearchResponseBuilderTest::testBuild
TypeError: Return value of Magento\Framework\Search\Response\QueryResponse::getTotal() must be of the type integer, null returned
/home/travis/build/magento/magento2/lib/internal/Magento/Framework/Search/Response/QueryResponse.php:81
/home/travis/build/magento/magento2/lib/internal/Magento/Framework/Search/SearchResponseBuilder.php:54
/home/travis/build/magento/magento2/lib/internal/Magento/Framework/Search/Test/Unit/SearchResponseBuilderTest.php:80
```

Could you please check these issues?

Thank you!

ronak2ram committed:
Update unit test

ronak2ram commented:
Hello @dmytro-ch
Thanks for the review of my PR. I updated the test case and resolved Travis error.
Can you please review again.

dmytro-ch commented:
@magento-engcom-team give me test instance

magento-engcom-team commented:
Hi @dmytro-ch. Thank you for your request. I'm working on Magento instance for you

magento-engcom-team commented:
Hi @dmytro-ch, here is your [new Magento instance](https://pr-21801.instances.magento-community.engineering).
Admin access: [https://pr-21801.instances.magento-community.engineering/admin](https://pr-21801.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

dmytro-ch committed:
Minor refactoring

magento-engcom-team commented:
Hi @dmytro-ch, thank you for the review.
[ENGCOM-4529](https://jira.corp.magento.com/browse/ENGCOM-4529) has been created to process this Pull Request

stoleksiy commented:
:heavy_check_mark: QA Passed

**Result:**

![after](https://user-images.githubusercontent.com/44244273/54665507-f637c500-4aef-11e9-9a40-09004dd8682c.png)

sidolov commented:
@ronak2ram please, apply latest fixes from the original PR: https://github.com/magento/magento2/pull/21713

sidolov reviewed (changes requested):
Fixes from the original PR is needed

ronak2ram commented:
@sidolov I did all the changes from #21713
Can you please help me to find remaining changes.

ronak2ram commented:
@sidolov I did all the change from this PR: #21713
Can you please help me to find the remaining changes.

sidolov commented:
Hi @ronak2ram , changeset from the https://github.com/magento/magento2/pull/21713 contains only 2 files, current PR contains 9 files. Please, make changes from the original PR.

ronak2ram commented:
Hi @sidolov 
I added new files change in this PR because someone added those changes in 2.3-develop(https://github.com/magento/magento2/commit/021a953482e06a622d1508fe35aa4e744d636493#diff-d3e053570d9a25c8e314852f808fec8a) branch so not need to add in 2.3-develop PR #21713.
But I see in 2.2-develop I do not find this commit https://github.com/magento/magento2/commit/021a953482e06a622d1508fe35aa4e744d636493#diff-d3e053570d9a25c8e314852f808fec8a
changes so I added that files and code in current PR which need for the get correct search count.

sidolov commented:
Hi @ronak2ram , unfortunately, such changes are not acceptable for 2.2 release line due to backward compatibility. Probably you may find a different solution for 2.2 without breaking changes.

ronak2ram commented:
okay @sidolov, we have to close this PR.

sidolov commented:
@ronak2ram thank you for collaboration!
